### PR TITLE
Add static-prefix option to mlflow ui and server

### PR DESCRIPTION
**Pull request by andrewmchen**
With this change, we start serving the UI from either `/` or `static-prefix` if it is set from the command line.

In addition, this change will change all existing routes to a hash based one: `http://localhost:5000/#/experiments/0/runs/a50907d0e0034aa790187f39b502de6c` for example

### CODE REVIEW

**Comment by mparkhe** [comment]
Can you add some mock tests for prefix and hash router

**Comment by andrewmchen** [comment]
@mparkhe sure I'll add some unit tests.
@mateiz It's much easier to implement the MLflow server interface with hash based routes. This way, we don't need catchall handlers to return `index.html` and we don't need complicated logic to return the CSS/JS (it can simply be a relative path to `index.html`.

**Comment by mateiz** [comment]
Ah, got it. Do hash-based routes make it harder to create bookmarks? If most browsers save them as bookmarks, then I think it's fine.

File: mlflow/cli.py [hunk 1]
```diff
@@ -121,14 +121,16 @@ def run(uri, entry_point, version, param_list, experiment_id, mode, cluster_spec
                    "other machines.")
 @click.option("--port", "-p", default=5000,
               help="The port to listen on (default: 5000).")
-def ui(file_store, host, port):
+@click.option("--static-prefix", default=None,
+              help="A prefix which will be prepended to the path of all static paths.")
```
Comment (mateiz) [inline mlflow/cli.py]: I don't think we want this option on `mlflow ui` necessarily. Just put it on `mlflow server`. The `ui` command will be for people who want to run the UI locally and I'd actually like to make it open a web browser too eventually, so we should use `server` for stuff people want to host behind a proxy and such.

Comment (mateiz) [reply mlflow/cli.py]: Also, I would make the option a bit more descriptive: "A URL prefix that will be used when serving static content (example: /mlflow)"

File: mlflow/server/__init__.py [hunk 2]
```diff
@@ -16,33 +17,41 @@
     app.add_url_rule(http_path, handler.__name__, handler, methods=methods)
 
 
+def _join_static_prefix(route):
```
Comment (mateiz) [inline mlflow/server/__init__.py]: I'd call this `_add_static_prefix`

**Comment by andrewmchen** [comment]
Python 2 tests blocked by this PR. https://github.com/databricks/mlflow/pull/196

**Pull request merged**